### PR TITLE
Recovery ID (v) calculations + Docs

### PR DESCRIPTION
**Pull request by timoth-y**
The recovery id bit here is agnostic to chain requirements (for Eth add 27 + chain_id*2 as of EIP-155). It is calculated for both plain ECDSA signatures with `Keypair.Sign` and for the decrypted adaptors with `EncryptedSignature.Decrypt`. The latter requires no additional information from the signer and can be calculated during decryption.

This PR also contains library documentation.

### CODE REVIEW

File: secp256k1/adaptor.go [hunk 1]
```diff
@@ -8,35 +8,51 @@ import (
...
 		s.Negate()
 	}
 
 	r := a.R.X
+	is_r_odd := byte(a.R.Y.IsOddBit())
 
 	return &Signature{
 		r: &r,
 		s: s,
+		v: is_r_odd ^ is_s_high,
```
Comment (noot) [inline secp256k1/adaptor.go:47]:
```suggestion
		v: is_r_odd ^ byte(s.IsOverHalfOrder()),
```

Comment (noot) [reply secp256k1/adaptor.go:47]: you could remove lines 31-34?

File: secp256k1/keys.go [hunk 2]
```diff
@@ -227,24 +266,31 @@ func sign(z, x *secp256k1.ModNScalar) (*Signature, error) {
...
+	if s.IsOverHalfOrder() {
+		is_s_high = 1
+		s.Negate()
+	}
+
 	return &Signature{
 		r: r_fp,
 		s: s,
-		v: 0, // TODO
+		v: is_r_odd ^ is_s_high,
```
Comment (noot) [inline secp256k1/keys.go:279]: similar comment here as before

File: secp256k1/keys.go [hunk 3]
```diff
@@ -227,24 +266,31 @@ func sign(z, x *secp256k1.ModNScalar) (*Signature, error) {
...
 }
 
-// Public ...
+// Public returns PublicKey component.
 func (kp *Keypair) Public() *PublicKey {
 	return kp.public
 }
 
-// Private ...
+// Private PrivateKey component.
```
Comment (noot) [inline secp256k1/keys.go:288]:
```suggestion
// Private returns PrivateKey component.
```

File: secp256k1/adaptor.go [hunk 4]
```diff
@@ -8,35 +8,51 @@ import (
...
+//
+// * Once you give the decrypted Signature to anyone who has seen EncryptedSignature,
+// they will be able to learn decryption key aka secret by calling RecoverFromAdaptorAndSignature.
 func (a *EncryptedSignature) Decrypt(sk *secp256k1.ModNScalar) (*Signature, error) {
 	y_inv := &secp256k1.ModNScalar{}
 	y_inv.InverseValNonConst(sk)
 	s := new(secp256k1.ModNScalar)
 	s.Mul2(a.s, y_inv)
 
+	is_s_high := byte(0)
```
[the remark this answers is not in the document]
Comment (timoth-y) [reply secp256k1/adaptor.go:31]: lol I know, other options seemed even worse, but today I come with `sIsHigh`...

File: secp256k1/adaptor_test.go [hunk 5]
```diff
@@ -89,10 +89,9 @@ func TestAdaptor_ValidPlain(t *testing.T) {
 	require.NoError(t, err)
 	require.True(t, secret.Equals(y))
 
-	// TODO: dleq check fails, probably due to hash issues
-	// ok, err = pubkey.VerifyAdaptor(messageHashStr, encryptionKey, adaptor)
-	// require.NoError(t, err)
-	// require.True(t, ok)
+	ok, err = pubkey.VerifyAdaptor(messageHashStr, encryptionKey, adaptor)
+	require.NoError(t, err)
+	require.True(t, ok)
```
[the remark this answers is not in the document]
Comment (timoth-y) [reply secp256k1/adaptor_test.go:94]: Not sure what was it tho, just never repeated after I migrated to the new lib, probably something minor that I fixed without giving it much though

File: secp256k1/keys.go (hunk 3 above)
Comment (timoth-y) [reply secp256k1/keys.go]: thx, missed it

**Pull request merged**